targets/TARGET_Cypress/TARGET_PSOC6/spi_api.c

```diff
@@ -0,0 +1,648 @@
+    if (tx_count < tx_length) {
+        tx_byte = *tx_buffer++;
+    }
+    // Send required number of bytes.
+    while (tx_count < trans_length) {
+        if (Cy_SCB_SPI_Write(obj->base, tx_byte)) {
+            ++tx_count;
+            // Get next byte to transfer.
+            if (tx_count < tx_length) {
+                tx_byte = *tx_buffer++;
+            } else {
+                tx_byte = (uint8_t)write_fill;
+            }
+        }
+        // If we have bytes to receive check the rx fifo.
+        if (rx_count < rx_length) {
+            if (Cy_SCB_SPI_GetNumInRxFifo(obj->base) > 0) {
+                *rx_buffer++ = (char)Cy_SCB_SPI_Read(obj->base);
+                ++rx_count;
+            }
+        }
+    }
+    // Wait for tx fifo to empty while reading received bytes.
+    while (!Cy_SCB_SPI_IsTxComplete(obj->base)) {
+        if ((rx_count < rx_length) && (Cy_SCB_SPI_GetNumInRxFifo(obj->base) > 0)) {
+            *rx_buffer++ = (char)Cy_SCB_SPI_Read(obj->base);
+            ++rx_count;
+        }
+    }
+    // Read any ramaining bytes from the fifo.
+    while (rx_count < rx_length) {
+        *rx_buffer++ = (char)Cy_SCB_SPI_Read(obj->base);
+    }
+    // Clean up if we have read less bytes than available.
+    Cy_SCB_SPI_ClearRxFifo(obj->base);
+    return trans_length;
+}
+
+/** Check if a value is available to read
+ *
+ * @param[in] obj The SPI peripheral to check
+ * @return non-zero if a value is available
+ */
+int  spi_slave_receive(spi_t *obj_in) {
+    spi_obj_t *obj = OBJ_P(obj_in);
+    if (obj->ms_mode == CY_SCB_SPI_SLAVE) {
+        return Cy_SCB_SPI_GetNumInRxFifo(obj->base);
+    } else {
+        return 0;
+    }
+}
+
+/** Get a received value out of the SPI receive buffer in slave mode
+ *
+ * Blocks until a value is available
+ * @param[in] obj The SPI peripheral to read
+ * @return The value received
+ */
+int  spi_slave_read(spi_t *obj_in)
+{
+    spi_obj_t *obj = OBJ_P(obj_in);
+
+    if (obj->ms_mode == CY_SCB_SPI_SLAVE) {
+        while (Cy_SCB_SPI_GetNumInRxFifo(obj->base) == 0) {
+            // Wait for data.
+        }
+        return Cy_SCB_SPI_GetNumInRxFifo(obj->base);
+    } else {
+        return (int)CY_SCB_SPI_RX_NO_DATA;
+    }
+}
+
+/** Write a value to the SPI peripheral in slave mode
+ *
+ * Blocks until the SPI peripheral can be written to
+ * @param[in] obj   The SPI peripheral to write
+ * @param[in] value The value to write
+ */
+void spi_slave_write(spi_t *obj_in, int value)
+{
+    spi_obj_t *obj = OBJ_P(obj_in);
+
+    if (obj->ms_mode == CY_SCB_SPI_SLAVE) {
+        while ((Cy_SCB_SPI_GetTxFifoStatus(obj->base) & CY_SCB_SPI_TX_NOT_FULL) == 0) {
+            // Wait for a place available in a fifo.
+        }
+        Cy_SCB_SPI_Write(obj->base, value);
+    }
+}
+
+/** Checks if the specified SPI peripheral is in use
+ *
+ * @param[in] obj The SPI peripheral to check
+ * @return non-zero if the peripheral is currently transmitting
+ */
+int  spi_busy(spi_t *obj) {
```

Comment:
Can you run astyle on the HAL implementation (`{` should be on the new line). Currently astyle check ignores target folder but will be enabled in the next release most likely.